src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let initialState = { ...state };
+
+  actions.forEach((action) => {
+    const { type, extraData, keysToRemove } = action;
+
+    if (type === 'addProperties') {
+      initialState = { ...initialState, ...extraData };
+    } else if (type === 'removeProperties') {
+      initialState = { ...initialState };
+
+      keysToRemove.forEach((key) => {
+        delete initialState[key];
+      });
+    } else if (type === 'clear') {
+      initialState = {};
+    } else {
+      throw new Error(`Unknown action type: ${type}`);
+    }
+
+    newState.push({ ...initialState });
```

Comment:
The push operation is repeated in every case of the action type handling. This could be done once after the `switch` statement to adhere to the DRY (Don't Repeat Yourself) principle.